pandas/formats/format.py

```diff
@@ -1951,7 +1951,12 @@ def _format_hierarchical_rows(self):
                 for spans, levels, labels in zip(level_lengths,
                                                  self.df.index.levels,
                                                  self.df.index.labels):
-                    values = levels.take(labels)
+
+                    if levels._can_hold_na:
+                        values = levels.take(labels, fill_value=True)
```

Comment:
@mpuels It does not work whether the multiindex level contains NaNs or not, but if it *can* contain NaNs.

```
In [4]: level = pd.Index(['a', 'b'])

In [5]: level._can_hold_na
Out[5]: True

In [6]: level.take([0,0,1], fill_value=True)
Out[6]: Index([u'a', u'a', u'b'], dtype='object')
```

But you are correct you need this conditional here. 
In principle, you could also make a one-liner of it by passing `levels._can_hold_na` to `allow_fill` in `take`